ml-agents-envs/mlagents/envs/base_env.py

```diff
@@ -0,0 +1,291 @@
+    reward: np.array
+    done: np.array
+    max_step: np.array
+    agent_id: np.array
+    action_mask: Optional[List[np.array]]
+
+    def get_agent_step_result(self, agent_id: int) -> StepResult:
+        """
+        returns the step result for a specific agent.
+        :param agent_id: The id of the agent
+        :returns: obs, reward, done, agent_id and optional action mask for a
+        specific agent
+        """
+        try:
+            agent_index = np.where(self.agent_id == agent_id)[0][0]
+        except IndexError as ie:
+            raise IndexError(
+                "agent_id {} is not present in the BatchedStepResult".format(agent_id)
+            ) from ie
+        agent_obs = []
+        for batched_obs in self.obs:
+            agent_obs.append(batched_obs[agent_index])
+        agent_mask = None
+        if self.action_mask is not None:
+            agent_mask = []
+            for mask in self.action_mask:
+                agent_mask.append(mask[0])
+        return StepResult(
+            obs=agent_obs,
+            reward=self.reward[agent_index],
+            done=self.done[agent_index],
+            max_step=self.max_step[agent_index],
+            agent_id=agent_id,
+            action_mask=agent_mask,
+        )
+
+    @staticmethod
+    def empty(spec):
+        """
+        Returns an empty BatchedStepResult.
+        :param spec: The AgentGroupSpec for the BatchedStepResult
+        """
+        obs = []
+        for shape in spec.observation_shapes:
+            obs += [np.zeros((0,) + shape, dtype=np.float32)]
+        return BatchedStepResult(
+            obs=obs,
+            reward=np.zeros(0, dtype=np.float32),
+            done=np.zeros(0, dtype=np.bool),
+            max_step=np.zeros(0, dtype=np.bool),
+            agent_id=np.zeros(0, dtype=np.int32),
+            action_mask=None,
+        )
+
+    def n_agents(self) -> int:
+        return len(self.agent_id)
+
+
+class ActionType(Enum):
+    DISCRETE = 0
+    CONTINUOUS = 1
+
+
+class AgentGroupSpec(NamedTuple):
+    """
+    A NamedTuple to containing information about the observations and actions
+    spaces for a group of Agents.
+     - observation_shapes is a List of Tuples of int : Each Tuple corresponds
+     to an observation's dimensions. The shape tuples have the same ordering as
+     the ordering of the BatchedStepResult and StepResult.
+     - action_type is the type of data of the action. it can be discrete or
+     continuous. If discrete, the action tensors are expected to be int32. If
+     continuous, the actions are expected to be float32.
+     - action_shape is:
+       - An int in continuous action space corresponding to the number of
+     floats that constitute the action.
+       - A Tuple of int in discrete action space where each int corresponds to
+       the number of discrete actions available to the agent.
+    """
+
+    observation_shapes: List[Tuple]
+    action_type: ActionType
+    action_shape: Any
```

Comment:
Why not `Union[int, Tuple[int, ...]]`?

Reply:
If it's because mypy doesn't like the `len(self.action_shape)`, I'd rather disable the checks on that line than drop the type hint entirely.

Reply:
How do I disable ?

Reply:
`# type: ignore` (preferably with an explanation of why you're disabling.

Reply:
reference `# type: ignore` on culprit line.